dpnp/backend/kernels/dpnp_krnl_random.cpp

```diff
@@ -533,6 +533,292 @@ void dpnp_rng_negative_binomial_c(void* result, const double a, const double p,
+            {
+                throw std::runtime_error("DPNP RNG Error: dpnp_rng_noncentral_chisquare_c() failed.");
+            }
+            /* squaring could result in an overflow */
+            vmdSqr(size, nvec, nvec, VML_HA);
+            vmdAdd(size, result1, nvec, result1, VML_HA);
+
+            mkl_free(nvec);
+        }
+        else if (df < 1)
+        {
+            /* noncentral_chisquare(df, nonc) ~ G( df/2 + Poisson(nonc/2), 2) */
+            double lambda;
+            int* pvec = nullptr;
+            pvec = (int*)mkl_malloc(size * sizeof(int), 64);
+            if (pvec == nullptr)
+            {
+                throw std::runtime_error("DPNP RNG Error: dpnp_rng_noncentral_chisquare_c() failed.");
+            }
+
+            lambda = 0.5 * nonc;
+            errcode = viRngPoisson(VSL_RNG_METHOD_POISSON_PTPE, get_rng_stream(), size, pvec, lambda);
+            if (errcode != VSL_STATUS_OK)
+            {
+                throw std::runtime_error("DPNP RNG Error: dpnp_rng_noncentral_chisquare_c() failed.");
+            }
+
+            shape = 0.5 * df;
+
+            if (0.125 * size > sqrt(lambda))
+            {
+                size_t* idx = nullptr;
+                double* tmp = nullptr;
+
+                idx = (size_t*)mkl_malloc(size * sizeof(size_t), 64);
+                if (idx == nullptr)
+                {
+                    throw std::runtime_error("DPNP RNG Error: dpnp_rng_noncentral_chisquare_c() failed.");
+                }
+
+                for (i = 0; i < size; i++)
+                    idx[i] = i;
+
+                std::sort(idx, idx + size, [pvec](size_t i1, size_t i2) { return pvec[i1] < pvec[i2]; });
+                /* idx now contains original indexes of ordered Poisson outputs */
+
+                /* allocate workspace to store samples of gamma, enough to hold entire output */
+                tmp = (double*)mkl_malloc(size * sizeof(double), 64);
+                if (tmp == nullptr)
+                {
+                    throw std::runtime_error("DPNP RNG Error: dpnp_rng_noncentral_chisquare_c() failed.");
+                }
+                for (i = 0; i < size;)
+                {
+                    size_t k, j;
+                    int cv = pvec[idx[i]];
+
+                    for (j = i + 1; (j < size) && (pvec[idx[j]] == cv); j++)
+                    {
+                    }
+                    // assert(j > i);
+                    if (j <= i)
+                    {
+                        throw std::runtime_error("DPNP RNG Error: dpnp_rng_noncentral_chisquare_c() failed.");
+                    }
+                    errcode = vdRngGamma(
+                        VSL_RNG_METHOD_GAMMA_GNORM_ACCURATE, get_rng_stream(), j - i, tmp, shape + cv, d_zero, d_two);
+                    if (errcode != VSL_STATUS_OK)
+                    {
+                        throw std::runtime_error("DPNP RNG Error: dpnp_rng_noncentral_chisquare_c() failed.");
```

Comment:
it is better to make one standalone function to check error code

Reply:
Yes, I have the same idea. Let me implement this feature on another PR. I will refactor all related code after.